File: src/main/java/base/mvp/BaseViewModel.java
package base.mvp;

import android.content.Context;
import android.databinding.BaseObservable;
import android.databinding.Observable;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class BaseViewModel extends BaseObservable implements BaseContract.ViewModel {

    @NonNull
    protected final Context mContext;

    private Map<Integer, List<Runnable>> mOnChangeObservers = new HashMap<>();


    public BaseViewModel(@NonNull Context context) {
        mContext = context;
        addOnPropertyChangedCallback(new OnPropertyChangedCallback() {
            @Override
            public void onPropertyChanged(Observable sender, int propertyId) {
                if (mOnChangeObservers.containsKey(propertyId)) {
                    for (Runnable action : mOnChangeObservers.get(propertyId)) {
                        action.run();
                    }
                }
            }
        });
    }

    @Override
    public void onPropertyChanged(int propertyId, Runnable action) {
        if (mOnChangeObservers.containsKey(propertyId)) {
            mOnChangeObservers.get(propertyId).add(action);
        } else {
            List<Runnable> observers = new ArrayList<>();
            observers.add(action);
            mOnChangeObservers.put(propertyId, observers);
        }
    }
}

File: src/main/java/base/mvp/BaseViewModel.kt
package base.mvp

import android.content.Context
import android.databinding.BaseObservable
import android.databinding.Observable
import android.util.SparseArray
import java.util.*

open class BaseViewModel(protected val mContext: Context) : BaseObservable(), BaseContract.ViewModel {

    private val mOnChangeObservers = SparseArray<MutableList<() -> Unit>>()


    init {
        addOnPropertyChangedCallback(object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable, propertyId: Int) {

                if (mOnChangeObservers[propertyId] != null) {
                    mOnChangeObservers[propertyId].forEach { it.invoke() }
                }
            }
        })
    }

    override fun onPropertyChanged(propertyId: Int, action: () -> Unit) {
        if (mOnChangeObservers[propertyId] != null) {
            mOnChangeObservers[propertyId].add(action)
        } else {
            val observers = ArrayList<() -> Unit>()
            observers.add(action)
            mOnChangeObservers.put(propertyId, observers)
        }
    }
}
